pygfx/renderers/wgpu/_update.py

```diff
@@ -50,28 +50,26 @@ def update_buffer(device, resource):
         resource._wgpu_usage |= wgpu.BufferUsage.COPY_DST
         buffer = device.create_buffer(size=resource.nbytes, usage=resource._wgpu_usage)
 
-    queue = device.queue
-    encoder = device.create_command_encoder()
+    resource._wgpu_buffer = resource.rev, buffer
+    if not pending_uploads:
+        return
 
     # Upload any pending data
     for offset, size in pending_uploads:
         subdata = resource._get_subdata(offset, size)
         # A: map the buffer, writes to it, then unmaps. But we don't offer a mapping API in wgpu-py
         # B: roll data in new buffer, copy from there to existing buffer
-        tmp_buffer = device.create_buffer_with_data(
-            data=subdata,
-            usage=wgpu.BufferUsage.COPY_SRC,
+        # tmp_buffer = device.create_buffer_with_data(
+        #     data=subdata,
+        #     usage=wgpu.BufferUsage.COPY_SRC,
+        # )
+        # boffset, bsize = bytes_per_item * offset, bytes_per_item * size
+        # encoder.copy_buffer_to_buffer(tmp_buffer, 0, buffer, boffset, bsize)
+        # C: using queue. This may be sugar for B, but it may also be optimized.
+        device.queue.write_buffer(
+            buffer, bytes_per_item * offset, subdata, 0, subdata.nbytes
         )
-        boffset, bsize = bytes_per_item * offset, bytes_per_item * size
-        encoder.copy_buffer_to_buffer(tmp_buffer, 0, buffer, boffset, bsize)
-        # C: using queue. This may be sugar for B, but it may also be optimized
-        # Unfortunately, this seems to crash the device :/
```

Comment:
It does not crash any more, it seems. Only tested on Linux yet.

Reply:
To clarify: there were 3 options here. This used option B, now I moved to option C, which was preferred anyway, but did not work previously.

Reply:
Tested with NVidia drivers on Linux. A confirmation on Windows would be nice.

Reply:
Which example should I run? :)

Reply:
Any example that uses a buffer would do. So e.g. anything with a mesh. If it does not crash or hang, I'm happy :)

Reply:
Here's what I get when I run `colormap_mesh.py`:

```
(.venv) PS C:\Users\kvang\dev\pygfx> python .\examples\colormap_mesh.py
Failed to parse WGSL code for Some(""): invalid field accessor `x`
thread '<unnamed>' panicked at 'Parsing(NagaParseError { shader_source: "        [[block]]\n        struct Struct_u_stdinfo {\n            cam_transform: mat4x4<f32>;\n            cam_transform_inv: mat4x4<f32>;\n            projection_transform: mat4x4<f32>;\n            projection_transform_inv: mat4x4<f32>;\n            physical_size: vec2<f32>;\n            logical_size: vec2<f32>;\n            flipped_winding: i32;\n        };\n\n        [[block]]\n        struct Struct_u_wobject {\n            world_transform: mat4x4<f32>;\n            world_transform_inv: mat4x4<f32>;\n            id: i32;\n        };\n\n        [[block]]\n        struct Struct_u_material {\n            color: vec4<f32>;\n            emissive_color: vec4<f32>;\n            clim: vec2<f32>;\n            opacity: f32;\n            wireframe: f32;\n            shininess: f32;\n        };\n\n        [[block]]\n        struct Buffer_u32 {\n            data: [[stride(4)]] array<u32>;\n        };\n\n        [[block]]\n        struct Buffer_f32 {\n            data: [[stride(4)]] array<f32>;\n        };\n\n        [[group(0), binding(0)]]\n        var<uniform> u_stdinfo: Struct_u_stdinfo;\n\n        [[group(0), binding(1)]]\n        var<uniform> u_wobject: Struct_u_wobject;\n\n        [[group(0), binding(2)]]\n        var<uniform> u_material: Struct_u_material;\n\n        [[group(1), binding(0)]]\n        var<storage, read> s_indices: Buffer_u32;\n        fn load_s_indices (i: i32) -> vec3<u32> { return vec3<u32>( s_indices.data[i * 3], s_indices.data[i * 3 + 1], s_indices.data[i * 3 + 2] ); }\n\n        [[group(1), binding(1)]]\n        var<storage, read> s_positions: Buffer_f32;\n        fn load_s_positions (i: i32) -> vec3<f32> { return vec3<f32>( s_positions.data[i * 3], s_positions.data[i * 3 + 1], s_positions.data[i * 3 + 2] ); }\n\n        [[group(1), binding(2)]]\n        var<storage, read> s_normals: Buffer_f32;\n        fn load_s_normals (i: i32) -> vec3<f32> { return vec3<f32>( s_normals.data[i * 3], s_normals.data[i * 3 + 1], s_normals.data[i * 3 + 2] ); }\n\n        [[group(1), binding(3)]]\n        var<storage, read> s_texcoords: Buffer_f32;\n        fn load_s_texcoords (i: i32) -> f32 { return s_texcoords.data[i]; }\n\n        [[group(1), binding(4)]]\n        var r_sampler: sampler;\n\n        [[group(1), binding(5)]]\n        var r_tex: texture_1d<f32>;\n            fn check_clipping_planes(world_pos: vec3<f32>) -> bool { return true; }\n            fn apply_clipping_planes(world_pos: vec3<f32>) { }\n            \n        fn ndc_to_world_pos(ndc_pos: vec4<f32>) -> vec3<f32> {\n            let ndc_to_world = u_stdinfo.cam_transform_inv * u_stdinfo.projection_transform_inv;\n            let world_pos = ndc_to_world * ndc_pos;\n            return world_pos.xyz / world_pos.w;\n        }\n        \n        var<private> p_pick_bits_used : i32 = 0;\n\n        fn pick_pack(value: u32, bits: i32) -> vec4<u32> {\n            // Utility to pack multiple values into a rgba16uint (64 bits available).\n            // Note that we store in a vec4<u32> but this gets written to a 4xu16.\n            // See #212 for details.\n            //\n            // Clip the given value\n            let v = min(value, u32(exp2(f32(bits))));\n            // Determine bit-shift for each component\n
  let shift = vec4<i32>(\n                p_pick_bits_used, p_pick_bits_used - 16, p_pick_bits_used - 32, p_pick_bits_used - 48,\n            );\n            // Prepare for next pack\n            p_pick_bits_used = p_pick_bits_used + bits;\n            // Apply the shift for each component\n            let vv = vec4<u32>(v);\n            let selector1 = vec4<bool>(shift[0] < 0, shift[1] < 0, shift[2] < 0, shift[3] < 0);\n            let pick_new = select( vv << vec4<u32>(shift) , vv >> vec4<u32>(-shift) , selector1 );\n            // Mask the components\n            let mask = vec4<u32>(65535u);\n            let selector2 = vec4<bool>( abs(shift[0]) < 32, abs(shift[1]) < 32, abs(shift[2]) < 32, abs(shift[3]) < 32 );\n            return select( vec4<u32>(0u) , pick_new & mask , selector2 );\n        }\n        \n        \n        var<private> p_fragment_color : vec4<f32> = vec4<f32>(0.0, 0.0, 0.0, 0.0);\n        var<private> p_fragment_depth : f32 = 1.1;\n\n        struct FragmentOutput {\n            [[location(0)]] color: vec4<f32>;\n            [[location(1)]] pick: vec4<u32>;\n        };\n        fn add_fragment(depth: f32, color: vec4<f32>) {\n            if (color.a >= 1.0 && depth < p_fragment_depth) {\n
p_fragment_color = vec4<f32>(color.rgb, 1.0);\n                p_fragment_depth = depth;\n            }\n        }\n        fn finalize_fragment() -> FragmentOutput {\n            if (p_fragment_depth > 1.0) { discard; }\n            var out : FragmentOutput;\n            out.color = p_fragment_color;\n            return out;\n        }\n        \n        \n\n        fn lighting_phong(\n            is_front: bool,\n            world_pos: vec3<f32>,\n            normal: vec3<f32>,\n
light: vec3<f32>,\n            view: vec3<f32>,\n            albeido: vec3<f32>,\n        ) -> vec3<f32> {\n            let light_color = vec3<f32>(1.0, 1.0, 1.0);\n\n            // Light parameters\n            let ambient_factor = 0.1;\n
  let diffuse_factor = 0.7;\n            let specular_factor = 0.3;\n            let shininess = u_material.shininess;\n\n            // Base vectors\n            var normal: vec3<f32> = normalize(normal);\n            let view = normalize(view);\n            let light = normalize(light);\n\n            // Maybe flip the normal - otherwise backfacing faces are not lit\n            // See pygfx/issues/#105 for details\n            normal = select(normal, -normal, is_front);\n\n            // Ambient\n            let ambient_color = light_color * ambient_factor;\n\n
 // Diffuse (blinn-phong reflection model)\n            let lambert_term = clamp(dot(light, normal), 0.0, 1.0);\n            let diffuse_color = diffuse_factor * light_color * lambert_term;\n\n            // Specular\n            let halfway = normalize(light + view);  // halfway vector\n            var specular_term = pow(clamp(dot(halfway,  normal), 0.0, 1.0), shininess);\n            specular_term = select(0.0, specular_term, shininess > 0.0);\n            let specular_color = specular_factor * specular_term * light_color;\n\n            // Emissive color is additive and unaffected by lights\n            let emissive_color = u_material.emissive_color.rgb;\n\n            // Put together\n            return albeido * (ambient_color + diffuse_color) + specular_color + emissive_color;\n        }\n\n        fn lighting_flat(\n            is_front: bool,\n            world_pos: vec3<f32>,\n            normal: vec3<f32>,\n            light: vec3<f32>,\n            view: vec3<f32>,\n            albeido: vec3<f32>,\n        ) -> vec3<f32> {\n\n            let u = dpdx(world_pos);\n            let v = dpdy(world_pos);\n            var normal = normalize(cross(u, v));\n\n            // The normal calculated above may not be oriented correctly.\n            // We have two flags: is_front and u_stdinfo.flipped_winding.\n            // Note that lighting_phong() also applies the is_front flag.\n            // Below code means: flip the normal if the XOR of these two flags is true.\n            normal = select(normal, -normal, (select(0, 1, is_front) + u_stdinfo.flipped_winding) == 1);\n\n            // The rest is the same as phong\n
  return lighting_phong(is_front, world_pos, normal, light, view, albeido);\n        }\n        struct VertexInput {\n            [[builtin(vertex_index)]] vertex_index : u32;\n        };\n\n        struct Varyings {\n            [[location(0)]] light : vec3<f32>;\n            [[location(1)]] normal : vec3<f32>;\n            [[location(2)]] pick_coords : vec3<f32>;\n            [[location(3)]] pick_id : u32;\n            [[location(4)]] pick_idx : u32;\n            [[location(5)]] texcoord : f32;\n            [[location(6)]] view : vec3<f32>;\n            [[location(7)]] world_pos : vec3<f32>;\n            [[builtin(position)]] position : vec4<f32>;\n        };\n\n        [[stage(vertex)]]\n        fn vs_main(in: VertexInput) -> Varyings {\n\n            // Select what face we're at\n            let index = i32(in.vertex_index);\n            let face_index = index / 3;\n            var sub_index = index % 3;\n\n            // If the camera flips a dimension, it flips the face winding.\n            // We can correct for this by adjusting the order (sub_index) here.\n            sub_index = select(sub_index, -1 * (sub_index - 1) + 1, u_stdinfo.flipped_winding > 0);\n\n            // Sample\n            let ii = load_s_indices(face_index);\n            let i0 = i32(ii[sub_index]);\n\n            // Get world transform\n                let world_transform = u_wobject.world_transform;\n            // Get vertex position\n            let raw_pos = load_s_positions(i0);\n            let world_pos = world_transform * vec4<f32>(raw_pos, 1.0);\n            var ndc_pos = u_stdinfo.projection_transform * u_stdinfo.cam_transform * world_pos;\n\n            // For the wireframe we also need the ndc_pos of the other vertices of this face\n            // Prepare output\n            var varyings: Varyings;\n\n            // Set position\n            varyings.world_pos = vec3<f32>(world_pos.xyz / world_pos.w);\n            varyings.position = vec4<f32>(ndc_pos.xyz, ndc_pos.w);\n\n            // Set texture coords\n            varyings.texcoord = f32(load_s_texcoords(i0));\n            // Set the normal\n            let raw_normal = load_s_normals(i0);\n            let world_pos_n = world_transform * vec4<f32>(raw_pos + raw_normal, 1.0);\n            let world_normal = normalize(world_pos_n - world_pos).xyz;\n            varyings.normal = vec3<f32>(world_normal);\n\n            // Vectors for lighting, all in world coordinates\n            let view_vec = normalize(ndc_to_world_pos(vec4<f32>(0.0, 0.0, 1.0, 1.0)));\n            varyings.view = vec3<f32>(view_vec);\n            varyings.light = vec3<f32>(view_vec);\n\n
// Set wireframe barycentric-like coordinates\n            // Set varyings for picking. We store the face_index, and 3 weights\n            // that indicate how close the fragment is to each vertex (barycentric\n            // coordinates). This allows the selection of the nearest vertex or edge.\n                let pick_id = u_wobject.id;\n            varyings.pick_id = u32(pick_id);\n            varyings.pick_idx = u32(face_index);\n            var arr_pick_coords = array<vec3<f32>, 3>(vec3<f32>(1.0, 0.0, 0.0), vec3<f32>(0.0, 1.0, 0.0), vec3<f32>(0.0, 0.0, 1.0));\n            varyings.pick_coords = vec3<f32>(arr_pick_coords[sub_index]);\n\n            return varyings;\n        }\n\n    \n\n        [[stage(fragment)]]\n        fn fs_main(varyings: Varyings, [[builtin(front_facing)]] is_front: bool) -> FragmentOutput {\n            var color_value: vec4<f32>;\n\n                        color_value = textureSample(r_tex, r_sampler, varyings.texcoord.x);\n                let albeido = (color_value.rgb - u_material.clim[0]) / (u_material.clim[1] - u_material.clim[0]);\n            // Lighting\n                let world_pos = varyings.world_pos;\n                let lit_color = lighting_phong(is_front, varyings.world_pos, varyings.normal, varyings.light, varyings.view, albeido);\n            let final_color = vec4<f32>(lit_color, color_value.a * u_material.opacity);\n\n            // Wrap up\n\n            apply_clipping_planes(varyings.world_pos);\n            add_fragment(varyings.position.z, final_color);\n            var out = finalize_fragment();\n\n            // The wobject-id must be 20 bits. In total it must not exceed 64 bits.\n            out.pick = (\n                pick_pack(varyings.pick_id, 20) +\n                pick_pack(varyings.pick_idx, 26) +\n                pick_pack(u32(varyings.pick_coords.x * 64.0), 6) +\n                pick_pack(u32(varyings.pick_coords.y * 64.0), 6) +\n                pick_pack(u32(varyings.pick_coords.z * 64.0), 6)\n            );\n            return out;\n        }\n", error: ParseError { message: "invalid field accessor `x`", labels: [(10829..10830, "invalid accessor")], notes: [] } })', src\device.rs:191:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Reply:
`mesh_slice.py` does work!